### PR TITLE
Quality tools

### PR DESCRIPTION
**Pull request by TheBonheurs**
These changes should bolster our programming quality by adding support within gradle for plugins like checkstyle, findbugs, pmd and lint. All have their own customizable rulesets and print their reports in the report directory (within app/build folders).

### CODE REVIEW

File: .idea/vcs.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ericcornelissen) [inline .idea/vcs.xml]: This file is in the `.gitignore`, please remove it manually from your branch

File: config/quality/checkstyle/checkstyle.xml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ericcornelissen) [inline config/quality/checkstyle/checkstyle.xml]: Why is there a `.../checkstyle.xml` and `...`checkstyle1.xml`?

File: config/quality.gradle [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        }
+    }
+
+    classpath = files()
+}
+
+task pmd(type: Pmd) {
+    ignoreFailures = true
+    // custom file, not sure if working yet
+	// ruleSetFiles = files("$configDir/pmd/pmd-ruleset.xml")
```
Comment (ericcornelissen) [inline config/quality.gradle]: Please remove commented code (and while you're at it, maybe remove the first (empty) line of this file 🙂)

File: config/quality/pmd/pmd-ruleset.xml [hunk 4]
```diff
@@ -0,0 +1,89 @@
+<?xml version="1.0"?>
+
+<ruleset name="PMD-Rules">
+  <description>
+	The rulesets we want to use in pmd. For a detailed description of pmd rulesets: http://pmd.sourceforge.net/rules/index.html
+  </description>
+  
+  
+  
+  <rule ref="rulesets/basic.xml/BooleanInstantiation">						<priority>5</priority></rule>
```
Comment (ericcornelissen) [inline config/quality/pmd/pmd-ruleset.xml]: If it is not to much effort, could you format this a little bit nicer 😄

**Comment by TheBonheurs** [comment]
Sharp finds! I'll make sure they are resolved asap. The pmd-ruleset file isn't needed anymore so I will just delete it instead of reformatting it. The checkstyle1.xml was an example document I found online for checkstyle rules but since it wasn't working and you can still find it online I will remove it aswell.

File: config/quality.gradle [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    include '**/*.java'
+    exclude '**/gen/**'
+    classpath = files()
+}
+
+task findbugs(type: FindBugs) {
+    ignoreFailures = false
+    effort = "max"
+    reportLevel = "high"
+    excludeFilter = file("$configDir/findbugs/findbugs-filter.xml")
```
Comment (ericcornelissen) [inline config/quality.gradle]: Do you have any idea what the difference between this line and line 30 (`exclude '**/gen/**') is? I found [this](http://findbugs.sourceforge.net/manual/filter.html) about `excludeFilters`, it seems it doesn't add anything alongside line 30...

File: config/quality.gradle [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    classes = files("${project.rootDir}/app/build/intermediates/classes")
```
Comment (ericcornelissen) [inline config/quality.gradle]: Does this line do anything for us? We don't have a folder `/app/build/...` in our project...

File: config/quality.gradle [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            destination "$reportsDir/findbugs/findbugs.html"
+        }
+    }
+
+    classpath = files()
+}
+
+task pmd(type: Pmd) {
+    ignoreFailures = true
+    ruleSets = [
```
Comment (ericcornelissen) [inline config/quality.gradle]: I think its better of the ruleset is in a seperate file (similar to Findbugs and Checkstyle)

**Comment by ericcornelissen** [comment]
About Travis-CI, it seems you didn't properly close some tags in the [checkstyle configuration .xml file](https://github.com/ericcornelissen/NervousFish/blob/c822c1ded2c78fa8e1d96dddc1c41861dbb451a5/config/quality/checkstyle/checkstyle.xml). The outermost `<module>` tag isn't closed, I suspect thats why Travis fails (see the error [here](https://travis-ci.org/ericcornelissen/NervousFish/builds/226588584#L2861)). It should be:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<module name="Checker">
  <module name="FileTabCharacter"/>
    <module name="TreeWalker">
      <module name="UnusedImports"/>
    </module>
  </module>
</module>
```

File: config/quality.gradle (hunk 6 above)
Comment (TheBonheurs) [inline config/quality.gradle]: We should after this pr ! This folder will contain things like reports from gradle plugins like findbugs etc.

File: config/quality.gradle (hunk 5 above)
Comment (TheBonheurs) [inline config/quality.gradle]: I have a separte file in that directory where I fileter some findbugs errors. The full path is config/quality/findbugs/findbugs-filter.xml

File: config/quality.gradle (hunk 7 above)
Comment (TheBonheurs) [inline config/quality.gradle]: I experimented with that earlier and it wasn't working too well and according to the newest stackoverflow answers regarding this config, using rulesets seems the right way to go http://stackoverflow.com/questions/20710704/gradles-pmd-plugin-what-are-acceptable-arguments

File: config/quality.gradle (hunk 5 above)
Comment (ericcornelissen) [inline config/quality.gradle]: 👍

File: config/quality.gradle (hunk 7 above)
Comment (ericcornelissen) [inline config/quality.gradle]: Once Checkstyle is working, we'll look at this 😉

Comment (ericcornelissen) [inline config/quality.gradle]: We'll create a separate issue to this so we can merge this working branch

File: config/quality.gradle [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    exclude '**/gen/**'
+    classpath = files()
+}
+
+checkstyle {
+    toolVersion = "6.0"
+}
+
+task findbugs(type: FindBugs) {
+    ignoreFailures = true
```
Comment (ericcornelissen) [inline config/quality.gradle]: Shouldn't this be false? I expect Travis to not fail on checkstyle errors using this config

File: config/quality.gradle [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+        html {
+            destination "$reportsDir/findbugs/findbugs.html"
+        }
+    }
+
+    classpath = files()
+}
+
+task pmd(type: Pmd) {
+    ignoreFailures = true
```
Comment (ericcornelissen) [inline config/quality.gradle]: Shouldn't this be false? I expect Travis to not fail on PMD errors using this config

File: config/quality.gradle [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+android {
+    lintOptions {
+        abortOnError true
+        xmlReport false
+        htmlReport true
+        lintConfig file("$configDir/lint/lint.xml")
+        htmlOutput file("$reportsDir/lint/lint-result.html")
+        xmlOutput file("$reportsDir/lint/lint-result.xml")
```
Comment (ericcornelissen) [inline config/quality.gradle]: If we keep xmlReport on false, this line is not needed

File: config/quality.gradle (hunk 7 above)
Comment (TheBonheurs) [inline config/quality.gradle]: Agreed !

File: config/quality.gradle (hunk 8 above)
Comment (TheBonheurs) [inline config/quality.gradle]: Yes I set them all to wrong value, good find

File: config/quality.gradle (hunk 10 above)
Comment (TheBonheurs) [inline config/quality.gradle]: Will delete it then 😄

File: config/quality.gradle (hunk 8 above)
Comment (ericcornelissen) [inline config/quality.gradle]: 👍

**Comment by TheBonheurs** [comment]
@ericcornelissen I think this pr is ready for merge, what do you think?

**Pull request merged**